### ISSUE TITLE
New Diagnoser for Thread Times / Blocking

### ISSUE THREAD

**Issue by feO2x** [dotnet/BenchmarkDotNet#1121]
I'm currently investigating the performance of FileStream in .NET / .NET Core and found that there are significant performance differences between Read and ReadAsync. While writing benchmarks for this, the following idea came to mind: is it possible to write a diagnoser that checks how long a thread was blocked during a benchmark?

This would be useful to measure the impact of thread blocking when performing I/O streaming with different buffer sizes. I have the following questions:

- Is this already possible via another diagnoser? Maybe ETW? On the fly, I couldn't find an [ETW event](https://docs.microsoft.com/en-us/dotnet/framework/performance/clr-etw-events) that shows when a thread blocks / unblocks.
- If this diagnoser has to be implemented, then could we use [the native GetThreadTimes function](https://docs.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-getthreadtimes)? Is the statement `blockedTime = totalTime - kernelTime - userTime` correct?

What's your opinion on the topic?

**Comment by AndreyAkinshin** [comment]
@feO2x thanks for the issue!
@adamsitnik is the best person who can answer this question, but he is on vacation right now. Let's wait until he returns.

**Comment by wojtpl2** [comment]
Maybe 'ConcurrencyVisualizerProfiler' will solve your problem? Adam wrote an article: https://adamsitnik.com/ConcurrencyVisualizer-Profiler/ 

![image](https://user-images.githubusercontent.com/17333903/55332679-29863680-5496-11e9-864b-aac0b8d455f2.png)

**Comment by adamsitnik** [comment]
Hi @feO2x 

Please excuse me for such a delay.

I think that this is out of scope of BenchmarkDotNet, you should use a profiler to get this information.

As @WojciechNagorski suggested, you should most probably use Concurrency Visualizer

Also, PerfView can help a bit: https://channel9.msdn.com/Series/PerfView-Tutorial/Tutorial-12-Wall-Clock-Time-Investigation-Basics
